File: app/src/main/java/framework/mvp/PresenterCompat.java
package framework.mvp;

import com.rxnetwork.manager.RxNetWorkListener;

/**
 * by y on 2017/2/24
 * <p>
 * <p>
 * M :  model
 * <p>
 * V ： View
 */

public abstract class PresenterCompat<M, V extends BaseView>
        implements RxNetWorkListener<M> {

    private V mView;

    public PresenterCompat(V mView) {
        this.mView = mView;
    }

    public V getView() {
        return mView;
    }

    @Override
    public void onNetWorkStart() {
        getView().showProgress();
    }

    @Override
    public void onNetWorkCompleted() {
        getView().hideProgress();
    }

    @Override
    public void onNetWorkError(Throwable e) {
        getView().hideProgress();
        getView().rxNetWorkError(e);
    }

    @Override
    public void onNetWorkSuccess(M data) {
        onNetWorkNext(data);
    }

    protected abstract void onNetWorkNext(M data);
}

File: app/src/main/java/framework/mvp/PresenterCompat.kt
package framework.mvp

import com.rxnetwork.manager.RxNetWorkListener

/**
 * by y on 2017/2/24
 *
 *
 *
 *
 * M :  model
 *
 *
 * V ： View
 */

abstract class PresenterCompat<M, out V : BaseView<*>>(val view: V) : RxNetWorkListener<M> {

    override fun onNetWorkStart() {
        view.showProgress()
    }

    override fun onNetWorkCompleted() {
        view.hideProgress()
    }

    override fun onNetWorkError(e: Throwable) {
        view.hideProgress()
        view.rxNetWorkError(e)
    }

    override fun onNetWorkSuccess(data: M) {
        onNetWorkNext(data)
    }

    protected abstract fun onNetWorkNext(data: M)
}

